// @netlify/zip-it-and-ship-it/dist/runtimes/node/bundlers/zisi/list_imports.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.listImports = void 0;
const esbuild = __importStar(require("@netlify/esbuild"));
const is_builtin_module_1 = __importDefault(require("is-builtin-module"));
const tmp_promise_1 = require("tmp-promise");
const consts_1 = require("../../../../utils/consts");
const fs_1 = require("../../../../utils/fs");
// Maximum number of log messages that an esbuild instance will produce. This
// limit is important to avoid out-of-memory errors due to too much data being
// sent in the Go<>Node IPC channel.
const ESBUILD_LOG_LIMIT = 10;
const getListImportsPlugin = ({ imports, path }) => ({
    name: 'list-imports',
    setup(build) {
        build.onResolve({ filter: /.*/ }, (args) => {
            const isEntryPoint = args.path === path;
            const isImport = !isEntryPoint && !(0, is_builtin_module_1.default)(args.path);
            if (isImport) {
                imports.add(args.path);
            }
            return {
                namespace: 'list-imports',
                external: isImport,
            };
        });
    },
});
const listImports = ({ functionName, path }) => __awaiter(void 0, void 0, void 0, function* () {
    // We're not interested in the output that esbuild generates, we're just
    // using it for its parsing capabilities in order to find import/require
    // statements. However, if we don't give esbuild a path in `outfile`, it
    // will pipe the output to stdout, which we also don't want. So we create
    // a temporary file to serve as the esbuild output and then get rid of it
    // when we're done.
    const targetPath = yield (0, tmp_promise_1.tmpName)();
    const imports = new Set();
    try {
        yield esbuild.build({
            bundle: true,
            entryPoints: [path],
            logLevel: 'error',
            logLimit: ESBUILD_LOG_LIMIT,
            outfile: targetPath,
            platform: 'node',
            plugins: [getListImportsPlugin({ imports, path })],
            target: 'esnext',
        });
    }
    catch (error) {
        error.customErrorInfo = {
            type: 'functionsBundling',
            location: { bundler: consts_1.JS_BUNDLER_ZISI, functionName, runtime: consts_1.RUNTIME_JS },
        };
        throw error;
    }
    finally {
        yield (0, fs_1.safeUnlink)(targetPath);
    }
    return [...imports];
});
exports.listImports = listImports;
